src/index.rs

```diff
@@ -1,38 +1,141 @@
 use rio_api::parser::TriplesParser;
 use rio_turtle::TurtleError;
-use std::{io::Write, path::Path};
+use serde::{Deserialize, Serialize};
+use smallvec::{smallvec, SmallVec};
+use std::{
+    collections::HashMap,
+    hash::{DefaultHasher, Hash, Hasher},
+    path::Path,
+};
 
 use crate::{
     io,
     rdf_types::{Triple, TripleView},
 };
 
-fn index_triple(t: Triple, out: &mut impl Write) {
+/// Stores a mapping from hashed instance uri to their types
+#[derive(Serialize, Deserialize)]
+pub struct TypeIndex {
+    pub types: Vec<String>,
+    map: HashMap<u64, SmallVec<[usize; 1]>>,
+}
+
+impl TypeIndex {
+    fn hash(&mut self, s: &str) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        s.hash(&mut hasher);
+        hasher.finish().to_le()
+    }
+
+    pub fn from_iter(type_map: impl Iterator<Item = (String, String)>) -> Self {
+        let vec: Vec<(String, String)> = type_map.collect();
```

Comment:
question: I dont understand the following:

1. We convert the iterator into a `Vec` with `collect` -> I think thats obsolete, because cant you just use the iterator directly, cause thats the whole point of passing a generic `T: Iterator<...` thing.

2. We do `vec.clone()` and then another `t.clone()` inside the map to only pass the type
3. This will get easier with `type_indices` as denoted above, by just calling `idx.insert` (?).